draft-ietf-mls-protocol.md

```diff
@@ -1190,6 +1206,10 @@ A step in this chain (the second subscript) is called a "generation".
 The details of application key derivation are described in the
 {{astree}} section below.
 
+[[ OPEN ISSUE: With the addition of Proposals, handshake encryption is now
+broken.  We need a framework where each sender has multiple keys, so we should
+probably just generate another AStree for handshake encryption. ]]
```

Comment:
It's a question of where you separate the handshake keys from the application keys.  You want them to be separate to simplify the analysis, so that the implication always goes Handshake => Application and not vice versa.

In the current draft, the nearest common ancestor is the epoch secret.  I think you're on the right track here, and a better design would be to use the AStree to fan out secrets per participant, then split out different keys for different uses at the leaves.

@beurdouche should probably opine here, but in any case, it seems like a follow-on PR, not something to address here.